plutus-use-cases/src/Plutus/Contracts/Uniswap.hs

```diff
@@ -0,0 +1,988 @@
+calculateRemoval inA inB liquidity diff = (f inA, f inB)
+  where
+    f :: Integer -> Integer
+    f x = x - divide (x * diff) liquidity
+
+data LiquidityPool = LiquidityPool
+    { lpCoinA :: Coin
+    , lpCoinB :: Coin
+    } deriving (Show, Generic, ToJSON, FromJSON, ToSchema)
+
+PlutusTx.unstableMakeIsData ''LiquidityPool
+PlutusTx.makeLift ''LiquidityPool
+
+instance Eq LiquidityPool where
+    {-# INLINABLE (==) #-}
+    x == y = (lpCoinA x == lpCoinA y && lpCoinB x == lpCoinB y) ||
+             (lpCoinA x == lpCoinB y && lpCoinB x == lpCoinA y)
+
+{-# INLINABLE hashLiquidityPool #-}
+hashLiquidityPool :: LiquidityPool -> ByteString
+hashLiquidityPool LiquidityPool{..} = sha2_256 $ concatenate (hashCoin c) (hashCoin d)
+  where
+    (c, d)
+        | lpCoinA `coinLT` lpCoinB = (lpCoinA, lpCoinB)
+        | otherwise                = (lpCoinB, lpCoinA)
+
+newtype Uniswap = Uniswap
+    { usCoin :: Coin
+    } deriving stock    (Show, Generic)
+      deriving anyclass (ToJSON, FromJSON, ToSchema)
+
+PlutusTx.makeLift ''Uniswap
+
+instance Prelude.Eq Uniswap where
+    u == v = usCoin u Prelude.== usCoin v
+
+instance Prelude.Ord Uniswap where
+    compare u v = Prelude.compare (usCoin u) (usCoin v)
+
+data UniswapAction = Create LiquidityPool | Close | Swap | Remove | Add
+    deriving Show
+
+PlutusTx.unstableMakeIsData ''UniswapAction
+PlutusTx.makeLift ''UniswapAction
+
+data UniswapDatum =
+      Factory [LiquidityPool]
+    | Pool LiquidityPool Integer
+    deriving stock (Show)
+
+PlutusTx.unstableMakeIsData ''UniswapDatum
+PlutusTx.makeLift ''UniswapDatum
+
+data Uniswapping
+instance Scripts.ScriptType Uniswapping where
+    type instance RedeemerType Uniswapping = UniswapAction
+    type instance DatumType Uniswapping = UniswapDatum
+
+
+{-# INLINABLE findOwnInput' #-}
+findOwnInput' :: ScriptContext -> TxInInfo
+findOwnInput' ctx = fromMaybe (error ()) (findOwnInput ctx)
+
+{-# INLINABLE valueWithin #-}
+valueWithin :: TxInInfo -> Value
+valueWithin = txOutValue . txInInfoResolved
+
+{-# INLINABLE checkSwap #-}
+checkSwap :: Integer -> Integer -> Integer -> Integer -> Bool
+checkSwap oldA oldB newA newB =
+    traceIfFalse "expected positive oldA" (oldA > 0) &&
+    traceIfFalse "expected positive oldB" (oldB > 0) &&
+    traceIfFalse "expected positive-newA" (newA > 0) &&
+    traceIfFalse "expected positive-newB" (newB > 0) &&
+    traceIfFalse "expected product to increase"
+        ((((newA * feeDen) - (inA * feeNum)) * ((newB * feeDen) - (inB * feeNum)))
+         >= (feeDen * feeDen * oldA * oldB))
+  where
+    inA, inB :: Integer
+    inA = max 0 $ newA - oldA
+    inB = max 0 $ newB - oldB
+
+-- (newA - fee * inA) * (newB - fee * inB) >= oldA * oldB
+-- (newA * feeDen - inA + feeNum) * (newB * feeDen - inB * feeNum)
+--     >= feeDen ^ 2 * oldA * oldB
+
+{-# INLINABLE validateSwap #-}
+validateSwap :: LiquidityPool -> Coin -> ScriptContext -> Bool
```

Comment:
a comment on each of the `validateXXX` functions would be nice, explaining what they are checking